README.md
# wordwraps-js

Apply common copy-editing fixes for text, specifically to control how lines wrap. We address widows at the end of a sentence, hyphens in the last word of a sentence, and patterns of word groups that should always remain together.

Assumes we're working with HTML content.

(This has a bias towards patterns seen in medical/health information.)

To control the wraps, we insert a `&nbsp;` entity between words.


# Installation

## With `npm`

If you're installing via `npm`... Ensure your project has an `.npmrc` file in the project root to tell `npm` where to find the package. Ensure the following line exists:

```
@kpander:registry=https://npm.pkg.github.com/
```

Then:

```
$ npm install @kpander/wordwraps-js
```

Then, annoyingly, you'll need to copy the files you need into your project's source folder, depending on where that is. E.g., assuming your source files are in `./src/js`:

```
$ cp "node_modules/@kpander/wordwraps-js/dist/Wordwraps.js" ./src/js
```


# Usage

## Example 1: Basic usage

```js
const input = `<p>Pretend this is a long sentence with a widow.</p>`;
console.log(Wordwraps.process(input));
```

```code
<p>Pretend this is a long sentence with a&nbsp;widow.</p>
```

# Features

  - The last two words in an element will be joined together
  - If the last word in an element has a hyphen, replace it with a non-breaking hyphen
  - Make certain word patterns non-breaking

## Word patterns

| Example      | Pattern       | Output   |
| :-           | :-            | :-       |
| `n = 25`     | "n = {###}"   | `n&nbsp;=&nbsp;25` |
| `5 days`     | "{###} {seconds\|minutes\|hours\|days\|weeks\|months\|years\|times\}" | `5&nbsp;days;` |
| `Phase 5`    | "Phase {###}  | `Phase&nbsp;5` |
| `Phase II`   | "Phase {I\|II\|III\|IV}" | `Phase&nbsp;II` |
| `Target 3`   | "Target {###} | `Target&nbsp;3` |
| `TARGET 3`   | "TARGET {###} | `TARGET&nbsp;3` |
| `P value`    | "P value"     | `P&nbsp;value`  |
| `<em>P</em> value` | `"<em>P</em> value"`     | `<em>P</em>&nbsp;value`  |
| `25% of`     | "{###}% of"   | `25%&nbsp;of` |
| `Day 5`      | "Day {###}"   | `Day&nbsp;5`  |
| `day 5`      | "day {###}"   | `day&nbsp;5`  |




# Developers

## Build distribution files

```bash
$ npm run build
```

This will build the distribution files in the `/dist/` folder. Run this before publishing a new release.


## Publishing a new version

This assumes you have an `.npmrc` file in the folder with a valid Github token for creating packages.

```bash
$ npm run build
$ npm publish
```


## Tests

You'll find Jest test cases in `./test`. To ensure all tests pass:

```bash
$ npm run test
```


# TODO

Note: This is definitely *not* a robust and production-ready library. It's a quick-and-dirty solution to a specific problem.

  - allow options for which elements are processed by each rule
  - allow providing custom word patterns to apply
  - allow defining *how* to make things non-breaking (`&nbsp;` vs wrapping in a `<span>`, etc.)


# Maintainer

  - Kendall Anderson (kpander@invisiblethreads.com)



CHANGELOG.md
# wordwraps-js Changelog

  - v2.0.1 (2023-09-05)
    - Maintenance: internal changes to prevent side effects

  - v2.0.0 (2023-09-02)
    - Breaking change: Rebuilt as the Wordwraps package

  - v1.1.0 (2015-05-01)
    - Feature: Adds additional word group pairs

  - v1.0.0 (2015-04-30)
    - Initial release as 'shrinkwrap' package


test/Wordwraps.hyphen.test.js
"use strict";
/*global describe, test, expect*/
/**
 * @file
 * Wordwraps.hyphen.test.js
 */

const Wordwraps = require("../src/Wordwraps");

const wrap = function(str) {
  return `
<!DOCTYPE html><html><head></head><body>
${str}
</body></html>
`.trim();
};

describe("Should replace non-breaking hyphens with entity", () => {

  test(
    `[Unit:NonBreakHyphen-001]
  Given
    - html string with no hyphen
  When
    - we process
  Then
    - should do nothing
`.trim(), async() => {
    // Given...
    const html = wrap("<p>this is my string with no hyphen</p>");
    const expected = html;

    // When...
    const result = Wordwraps._nonBreakingHyphen(html);

    // Then...
    expect(result).toEqual(expected);
  });

  test(
    `[Unit:NonBreakHyphen-002]
  Given
    - html string with an entity hyphen
  When
    - we process
  Then
    - should do nothing
`.trim(), async() => {
    // Given...
    const html = wrap("<p>this is my string with&#8209;hyphen</p>");
    const expected = html;

    // When...
    const result = Wordwraps._nonBreakingHyphen(html);

    // Then...
    expect(result).toEqual(expected);
  });

  test(
    `[Unit:NonBreakHyphen-003]
  Given
    - html string with a hyphen that isn't in the last word
  When
    - we process
  Then
    - should do nothing
`.trim(), async() => {
    // Given...
    const html1 = wrap("<p>this ibs-d is here</p>");
    const expected1 = html1;
    const html2 = wrap("<p>this ibs-d here</p>");
    const expected2 = html2;

    // When...
    const result1 = Wordwraps._nonBreakingHyphen(html1);
    const result2 = Wordwraps._nonBreakingHyphen(html2);

    // Then...
    expect(result1).toEqual(expected1);
    expect(result2).toEqual(expected2);
  });

  test(
    `[Unit:NonBreakHyphen-004]
  Given
    - html string with a hyphen in the last word
  When
    - we process
  Then
    - replace the hyphen with a nbsp entity hyphen
`.trim(), async() => {
    // Given...
    const html = wrap("<p>this is my string with ibs-d</p>");
    const expected = wrap("<p>this is my string with ibs&#8209;d</p>");

    // When...
    const result = Wordwraps._nonBreakingHyphen(html);

    // Then...
    expect(result).toEqual(expected);
  });

});



test/Wordwraps.widows.test.js
"use strict";
/*global describe, test, expect*/
/**
 * @file
 * Wordwraps.widows.test.js
 */

const Wordwraps = require("../src/Wordwraps");

const wrap = function(str) {
  return `
<!DOCTYPE html><html><head></head><body>
${str}
</body></html>
`.trim();
};

describe("Join the last 2 words in a paragraph", () => {

  test(
    `[JoinLastTwo-001]
  Given
    - html string
  When
    - we process
  Then
    - the last 2 words in the paragraph are combined with &nbsp;
`.trim(), async() => {
    // Given...
    const html = wrap("<p>this is my paragraph with last words</p>");
    const expected = wrap("<p>this is my paragraph with last&nbsp;words</p>");

    // When...
    const result = Wordwraps.process(html);

    // Then...
    expect(result).toEqual(expected);
  });

  /*
  test(
    `[JoinLastTwo-002]
  Given
    - html string with <sup> at end of sentence
  When
    - we process
  Then
    - should not join the last 2 words when either of the last 2 have markup
`.trim(), async() => {
    // Given...
    const html1 = wrap(`<p>these are my <span class="something">words</span></p>`);
    const expected1 = html1;
    const html2 = wrap(`<p>these are <span class="something">my </span>words</p>`);
    const expected2 = html2;

    // When...
    const result1 = Wordwraps.process(html1);
    const result2 = Wordwraps.process(html2);

    // Then...
    expect(result1).toEqual(expected1);
    expect(result2).toEqual(expected2);
  });
  */

  test(
    `[JoinLastTwo-003]
  Given
    - html string with <sup> at end of sentence
  When
    - we process
  Then
    - the last 2 words should be joined even if the last has a <sup> element
`.trim(), async() => {
    // Given...
    const html = wrap("<p>this is my paragraph with last words<sup>1</sup></p>");
    const expected = wrap("<p>this is my paragraph with last&nbsp;words<sup>1</sup></p>");

    // When...
    const result = Wordwraps.process(html);

    // Then...
    expect(result).toEqual(expected);
  });

  test(
    `[JoinLastTwo-004]
  Given
    - html string with a single word
  When
    - we process
  Then
    - we do nothing when there is only 1 word
`.trim(), async() => {
    // Given...
    const html = wrap("<p>thislongword</p>");
    const expected = html;

    // When...
    const result = Wordwraps.process(html);

    // Then...
    expect(result).toEqual(expected);
  });

  test(
    `[JoinLastTwo-005]
  Given
    - html string where the last 2 words are already joined with &nbsp;
  When
    - we process
  Then
    - we should leave the existing &nbsp; alone
`.trim(), async() => {
    // Given...
    const html = wrap("<p>this long sentence with last two&nbsp;joined</p>");
    const expected = html;

    // When...
    const result = Wordwraps.process(html);

    // Then...
    expect(result).toEqual(expected);
  });

});



test/Wordwraps.integration.test.js
"use strict";
/*global describe, test, expect*/
/**
 * @file
 * Wordwraps.integration.test.js
 */

const jsdom = require("jsdom");
const { JSDOM } = jsdom;

const Wordwraps = require("../src/Wordwraps");
const fs = require("fs");
const path = require("path");

const getFileContents = function(filename) {
  const file = path.join(__dirname, "artifacts", filename);
  return fs.readFileSync(file, "utf8").trim();
}

// Construct a js dom object for the given html string.
const dom = function(html) {
  return new JSDOM(html);
};

// Return the inner html for a dom element with the given id.
const innerHTML = function(dom, id) {
  return dom.window.document.getElementById(id).innerHTML;
}

describe("Full integration tests", () => {

  test(
    `[Integration-001]
  Given
    - html string
  When
    - we process
  Then
    - all expected changes are made
`.trim(), async() => {
    // Given...
    const html = getFileContents("full.source.html");
    const expected = getFileContents("full.expected.html");

    // When...
    const result = Wordwraps.process(html);
    const dom_result = dom(result);
    const dom_expected = dom(expected);

    // Then...
    // Go through all the ext[#] ids and make sure they pass.
    for (let i = 1; i < 18; i++) {
      const id = `exp${i}`;
      expect(innerHTML(dom_result, id).length).toBeGreaterThan(0);
      expect(innerHTML(dom_result, id)).toEqual(innerHTML(dom_expected, id));
    }
  });

});



src/Wordwraps.js
"use strict";
/**
 * @file
 * Wordwraps.js
 */

const jsdom = require("jsdom");
const { JSDOM } = jsdom;

const nobreakspace = '&nbsp;';
const nobreakhyphen = '&#8209;';

class Wordwraps {

  static process(html, srcOptions = {}) {
    const defaults = {
      widows: "h1, h2, h3, h4, h5, h6, p, li, th, td, dt, dd, a",
      hyphens: "h1, h2, h3, h4, h5, h6, o, li, th, td, dt, dd, a, span",
      groups: "h1, h2, h3, h4, h5, h6, o, li, th, td, dt, dd, a, span",
    };

    let options = { ...srcOptions };
    options = { ...defaults, ...options };

    const dom = new JSDOM(html);

    dom.window.document.querySelectorAll(options.widows).forEach(el => {
      el.innerHTML = Wordwraps._joinLastTwoWords(el.innerHTML);
    });

    dom.window.document.querySelectorAll(options.groups).forEach(el => {
      el.innerHTML = Wordwraps._nonBreakingHyphen(el.innerHTML);
    });

    dom.window.document.querySelectorAll(options.groups).forEach(el => {
      el.innerHTML = Wordwraps._joinWordGroups(el.innerHTML);
    });

    return dom.serialize();
  }

  static _joinLastTwoWords(html) {
    //console.log("\n-----------------------\ngiven html:", html);
    var words = html.split(' ');
    if (words.length < 2) {
      // Only 1 word, or an empty tag. Ignore it.
      //console.log("ignoring: < 2 words");
      return html;
    }

    const last1 = words.pop();
    const last2 = words.pop();

    if (last1.match(/&nbsp;/)) {
      // The last word already contains a non-breaking space, which means we
      // can't easily determine if it must be there, or if it's the result of
      // previously having run this filter. Ignore it.
      //console.log("last word already has nbsp:", last1);
      return html;
    }

    // Join the last 2 words in the element with a non-breaking space.
    words.push(last2 + nobreakspace + last1);
    html = words.join(' ');

    return html;
  }


  /**
   * If the last word in an element has a hyphen, replace with a non-breaking
   * hyphen.
   */
  static _nonBreakingHyphen(html) {
    let words = html.split(" ");
    let last = words.pop();

    if (Wordwraps._hasMarkup(last)) {
      // Last word has markup. Assume everything to the left of the first '<'
      // is the word, and everything to the right is markup. Yes, this is brittle.
      let parts = last.split("<");
      if (!parts[0].match(/-/)) return html;

      parts[0] = parts[0].replace(/-/g, nobreakhyphen);
      last = parts.join("<");
    } else if (last.match(/-/)) {
      last = last.replace(/-/g, nobreakhyphen);
    }

    words.push(last);
    html = words.join(' ');
    return html;
  }


  /**
   * Determine if the given string contains html markup.
   */
  static _hasMarkup(html) {
    if (html.match(/<\/?[a-z].*>/i)) {
      return true;
    }
    else {
      return false;
    }
  }

  /**
   * Join all of the defined word groups together with &nbsp;.
   */
  static _joinWordGroups(html) {
    var patterns = Wordwraps._getGroupPatterns();

    for (var i = 0; i < patterns.length; i++) {
      html = Wordwraps._applyRegex(html, patterns[i][1], patterns[i][2]);
    }

    return html;
  }


  static _applyRegex(html, pattern_search, pattern_replace) {
    return html.replace(pattern_search, pattern_replace);
  }


  static _getGroupPatterns() {
    var patterns = [];

    // Replace the spaces with &nbsp; in this pattern:
    //   "n = 25" (any number)
    patterns.push([
      "n = 25",
      /\bn\s+=\s+(\d+)/g,
      "n&nbsp;=&nbsp;$1"
    ]);


    // Replace the spaces with &nbsp; in this pattern:
    //   "25 minutes" (any number, and can be 'minutes' or 'minute')
    //   "10 hours"
    //   "3 days"
    //   "2 weeks"
    //   "1 month"
    //   "4 years"
    //   "1 time"
    patterns.push([
      "n minutes, n hours, n weeks",
      /\b(\d+)\s+(seconds?|minutes?|hours?|days?|weeks?|months?|years?|times?)\b/g,
      "$1&nbsp;$2"
    ]);


    // Replace the spaces with &nbsp; in this pattern:
    //   "Phase 3" (any number)
    //   "phase 3" (lowercase)
    //   "Phase III" (from I to IV)
    patterns.push([
      "Phase n",
      /\b(Phase|phase)\s+([\d]+|I|II|III|IV)\b/g,
      "$1&nbsp;$2"
    ]);


    // Replace the spaces with &nbsp; in this pattern:
    //   "Target 1" (any number)
    //   "TARGET 1" (uppercase)
    patterns.push([
      "Target n",
      /\b(Target|TARGET)\s+(\d+)\b/g,
      "$1&nbsp;$2"
    ]);


    // Replace the spaces with &nbsp; in this pattern:
    //   "<em>P</em> value"
    patterns.push([
      "<em>P</em> value",
      /<em>P<\/em> value/g,
      "<em>P</em>&nbsp;value"
    ]);
    patterns.push([
      "P value",
      /\bP value\b/g,
      "P&nbsp;value"
    ]);


    // Replace the spaces with &nbsp; in this pattern:
    //   "25% of" (any number)
    patterns.push([
      "25% of",
      /\b([0-9]+%)\sof\b/g,
      "$1&nbsp;of"
    ]);


    // Replace the spaces with &nbsp; in this pattern:
    //   "day 1" (any number)
    patterns.push([
      "day 2",
      /\b(Day|day)\s+([0-9]+)/g,
      "$1&nbsp;$2"
    ]);

    // @todo
    //   "GI tract"
    //   "\d mg" -- measurements?
    //   "2-week" "10-week" = wrap with span
    //   "x ≥ y" - nbsps
    //   "P=num.num" = wrap with span

    return patterns;
  }

}

module.exports = Wordwraps;
